File: packages/core_crypto/ios/core_crypto/Sources/core_crypto/WalletCoreBridge.swift
import Foundation
import WalletCore

/// Thin wrapper over Trust Wallet Core. Everything that touches private key
/// material lives here; buffers are zeroed immediately after use
/// (detailed-design.md §2.3).
enum WalletCoreBridge {
  enum BridgeError: Error { case invalidMnemonic, invalidInput, signFailed }

  static func generateMnemonic(strength: Int32) throws -> String {
    guard strength == 128 || strength == 192 || strength == 256,
      let wallet = HDWallet(strength: strength, passphrase: "")
    else { throw BridgeError.invalidInput }
    defer { /* HDWallet zeroizes on dealloc */ }
    return wallet.mnemonic
  }

  static func isValidMnemonic(_ mnemonic: String) -> Bool {
    Mnemonic.isValid(mnemonic: mnemonic)
  }

  static func isValidWord(_ word: String) -> Bool {
    Mnemonic.isValidWord(word: word)
  }

  static func suggest(_ prefix: String) -> [String] {
    Mnemonic.suggest(prefix: prefix).split(separator: " ").map(String.init)
  }

  /// Converts a mnemonic to entropy for storage (never store the string).
  static func entropy(from mnemonic: String) throws -> Data {
    guard isValidMnemonic(mnemonic) else { throw BridgeError.invalidMnemonic }
    guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: "") else {
      throw BridgeError.invalidMnemonic
    }
    return wallet.entropy
  }

  private static func wallet(fromEntropy entropy: Data) throws -> HDWallet {
    guard let wallet = HDWallet(entropy: entropy, passphrase: "") else {
      throw BridgeError.invalidMnemonic
    }
    return wallet
  }

  static func addresses(fromEntropy entropy: Data) throws -> [String: String] {
    let wallet = try wallet(fromEntropy: entropy)
    let evm = wallet.getAddressForCoin(coin: .ethereum)
    return [
      "eth": evm,
      "polygon": wallet.getAddressForCoin(coin: .polygon),
      "base": evm,
      "arbitrum": evm,
      "avalanche": evm,
      "bnb": evm,
      "tron": wallet.getAddressForCoin(coin: .tron),
      "solana": wallet.getAddressForCoin(coin: .solana),
    ]
  }

  static func publicKeys(fromEntropy entropy: Data) throws -> [String: Data] {
    let wallet = try wallet(fromEntropy: entropy)
    let evm = wallet.getKeyForCoin(coin: .ethereum)
      .getPublicKeySecp256k1(compressed: false).data
    let tron = wallet.getKeyForCoin(coin: .tron)
      .getPublicKeySecp256k1(compressed: false).data
    let solana = wallet.getKeyForCoin(coin: .solana)
      .getPublicKeyEd25519().data
    return [
      "eth": evm,
      "polygon": evm,
      "base": evm,
      "arbitrum": evm,
      "avalanche": evm,
      "bnb": evm,
      "tron": tron,
      "solana": solana,
    ]
  }

  /// Derives the three unique account keys used by KT Wallet. All EVM
  /// networks intentionally share the Ethereum account path/address.
  /// Returned buffers are owned by the caller and must be wiped.
  static func privateKeys(fromEntropy entropy: Data) throws -> [String: Data] {
    let wallet = try wallet(fromEntropy: entropy)
    let keys = [
      "evm": wallet.getKeyForCoin(coin: .ethereum).data,
      "tron": wallet.getKeyForCoin(coin: .tron).data,
      "solana": wallet.getKeyForCoin(coin: .solana).data,
    ]
    guard keys.values.allSatisfy({ $0.count == 32 }) else {
      throw BridgeError.invalidInput
    }
    return keys
  }

  static func encodePrivateKey(_ key: Data, coin: String) throws -> String {
    guard key.count == 32 else { throw BridgeError.invalidInput }
    switch coin {
    case "solana": return Base58.encodeNoCheck(data: key)
    case "eth", "polygon", "base", "arbitrum", "avalanche", "bnb", "tron":
      return key.hexString
    default: throw BridgeError.invalidInput
    }
  }

  private static func coinType(_ coin: String) -> CoinType? {
    switch coin {
    case "eth": return .ethereum
    case "polygon": return .polygon
    case "base", "arbitrum", "avalanche", "bnb": return .ethereum
    case "tron": return .tron
    case "solana": return .solana
    default: return nil
    }
  }

  static func sign(entropy: Data, coin: String, signingInput: Data) throws
    -> (signedTx: Data, txHash: String)
  {
    guard let type = coinType(coin) else { throw BridgeError.invalidInput }
    if type == .tron { return try signTron(entropy: entropy, rawData: signingInput) }
    if type == .solana { return try signSolana(entropy: entropy, message: signingInput) }
    return try signEvm(entropy: entropy, coin: type, encoded: signingInput)
  }

  private static func signEvm(entropy: Data, coin: CoinType, encoded: Data) throws
    -> (signedTx: Data, txHash: String)
  {
    let tx = try decodeEip1559(encoded)
    let key = try wallet(fromEntropy: entropy).getKeyForCoin(coin: coin)
    let input = EthereumSigningInput.with {
      $0.chainID = tx.chainID
      $0.nonce = tx.nonce
      $0.txMode = .enveloped
      $0.maxInclusionFeePerGas = tx.maxPriorityFeePerGas
      $0.maxFeePerGas = tx.maxFeePerGas
      $0.gasLimit = tx.gasLimit
      $0.toAddress = "0x" + tx.to.hexString
      $0.privateKey = key.data
      $0.transaction = EthereumTransaction.with {
        if tx.data.isEmpty {
          $0.transfer = EthereumTransaction.Transfer.with { $0.amount = tx.value }
        } else {
          $0.contractGeneric = EthereumTransaction.ContractGeneric.with {
            $0.amount = tx.value
            $0.data = tx.data
          }
        }
      }
    }
    let output: EthereumSigningOutput = AnySigner.sign(input: input, coin: coin)
    guard output.error == .ok, !output.encoded.isEmpty else {
      throw BridgeError.signFailed
    }
    let hash = Hash.keccak256(data: output.encoded)
    return (output.encoded, "0x" + hash.hexString)
  }

  /// Signs canonical TRON `Transaction.raw` bytes and emits the broadcasthex
  /// payload: `{"transaction": "<hex>", "txID": "<hex>"}`.
  ///
  /// `transaction` is the full signed `Transaction` protobuf (`raw_data` =
  /// field 1, `signature` = repeated field 2), which is what
  /// `/wallet/broadcasthex` takes. Emitting only `raw_data_hex` + `signature`
  /// is NOT broadcastable: `/wallet/broadcasttransaction` dereferences the
  /// `raw_data` object and answers a bare NullPointerException without it.
  private static func signTron(entropy: Data, rawData: Data) throws
    -> (signedTx: Data, txHash: String)
  {
    guard !rawData.isEmpty else { throw BridgeError.invalidInput }
    let key = try wallet(fromEntropy: entropy).getKeyForCoin(coin: .tron)
    let txID = Hash.sha256(data: rawData)
    guard let signature = key.sign(digest: txID, curve: .secp256k1),
          signature.count == 65 else { throw BridgeError.signFailed }
    let json: [String: Any] = [
      "transaction": tronSignedTransaction(rawData: rawData, signature: signature).hexString,
      "txID": txID.hexString,
    ]
    return (
      try JSONSerialization.data(withJSONObject: json, options: []),
      txID.hexString
    )
  }

  /// Serializes `Transaction { raw_data = 1; repeated signature = 2; }` —
  /// each field is a length-delimited (wire type 2) tag byte, a varint
  /// length, then the payload.
  private static func tronSignedTransaction(rawData: Data, signature: Data) -> Data {
    var out = Data([0x0a])
    out.append(protobufVarint(rawData.count))
    out.append(rawData)
    out.append(0x12)
    out.append(protobufVarint(signature.count))
    out.append(signature)
    return out
  }

  /// Base-128 varint, low group first, continuation bit on every group but the last.
  private static func protobufVarint(_ value: Int) -> Data {
    var remaining = value
    var out = Data()
    repeat {
      let group = UInt8(remaining & 0x7f)
      remaining >>= 7
      out.append(remaining != 0 ? group | 0x80 : group)
    } while remaining != 0
    return out
  }

  private static func signSolana(entropy: Data, message: Data) throws
    -> (signedTx: Data, txHash: String)
  {
    guard !message.isEmpty else { throw BridgeError.invalidInput }
    let key = try wallet(fromEntropy: entropy).getKeyForCoin(coin: .solana)
    guard let signature = key.sign(digest: message, curve: .ed25519),
          signature.count == 64 else { throw BridgeError.signFailed }
    return (
      Data([1]) + signature + message,
      Base58.encodeNoCheck(data: signature)
    )
  }

  private struct Eip1559Fields {
    let chainID, nonce, maxPriorityFeePerGas, maxFeePerGas: Data
    let gasLimit, to, value, data: Data
  }

  private struct RlpItem {
    let bytes: Data
    let isList: Bool
    let next: Int
  }

  private static func decodeEip1559(_ input: Data) throws -> Eip1559Fields {
    guard input.count > 2, input[0] == 0x02 else {
      throw BridgeError.invalidInput
    }
    let outer = try readRlp(input, at: 1)
    guard outer.isList, outer.next == input.count else {
      throw BridgeError.invalidInput
    }
    var fields: [RlpItem] = []
    var offset = 0
    while offset < outer.bytes.count {
      let item = try readRlp(outer.bytes, at: offset)
      fields.append(item)
      offset = item.next
    }
    guard fields.count == 9,
          !fields[0...7].contains(where: { $0.isList }),
          fields[8].isList, fields[8].bytes.isEmpty,
          fields[5].bytes.count == 20 else {
      throw BridgeError.invalidInput
    }
    return Eip1559Fields(
      chainID: fields[0].bytes, nonce: fields[1].bytes,
      maxPriorityFeePerGas: fields[2].bytes, maxFeePerGas: fields[3].bytes,
      gasLimit: fields[4].bytes, to: fields[5].bytes,
      value: fields[6].bytes, data: fields[7].bytes)
  }

  private static func readRlp(_ data: Data, at offset: Int) throws -> RlpItem {
    guard offset < data.count else { throw BridgeError.invalidInput }
    let prefix = Int(data[offset])
    if prefix <= 0x7f {
      return RlpItem(bytes: Data([data[offset]]), isList: false, next: offset + 1)
    }
    if prefix <= 0xb7 {
      let length = prefix - 0x80
      return try rlpPayload(data, offset, 1, length, false)
    }
    if prefix <= 0xbf {
      let lengthOfLength = prefix - 0xb7
      let length = try rlpLength(data, offset + 1, lengthOfLength)
      return try rlpPayload(data, offset, 1 + lengthOfLength, length, false)
    }
    if prefix <= 0xf7 {
      let length = prefix - 0xc0
      return try rlpPayload(data, offset, 1, length, true)
    }
    let lengthOfLength = prefix - 0xf7
    let length = try rlpLength(data, offset + 1, lengthOfLength)
    return try rlpPayload(data, offset, 1 + lengthOfLength, length, true)
  }

  private static func rlpLength(_ data: Data, _ start: Int, _ count: Int) throws -> Int {
    guard count > 0, count <= 4, start + count <= data.count else {
      throw BridgeError.invalidInput
    }
    var value = 0
    for byte in data[start..<(start + count)] { value = (value << 8) | Int(byte) }
    return value
  }

  private static func rlpPayload(
    _ data: Data, _ offset: Int, _ header: Int, _ length: Int, _ isList: Bool
  ) throws -> RlpItem {
    let start = offset + header
    let end = start + length
    guard length >= 0, end <= data.count else { throw BridgeError.invalidInput }
    return RlpItem(bytes: data.subdata(in: start..<end), isList: isList, next: end)
  }
}

File: packages/core_crypto/android/src/walletcore/kotlin/com/ktwallet/core_crypto/WalletCoreBridge.kt
package com.ktwallet.core_crypto

import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.Curve
import wallet.core.jni.Base58
import wallet.core.jni.HDWallet
import wallet.core.jni.Hash
import wallet.core.jni.Mnemonic
import wallet.core.jni.proto.Ethereum

/**
 * Trust Wallet Core wrapper. All private-key handling lives here; sensitive
 * byte arrays are zeroed after use (detailed-design.md §2.3).
 */
object WalletCoreBridge {
    init {
        System.loadLibrary("TrustWalletCore")
    }

    class InvalidMnemonicException : Exception("invalid mnemonic")
    class InvalidInputException : Exception("invalid signing input")
    class SignFailedException : Exception("signing failed")
    // Kept in the common bridge surface so CoreCryptoPlugin can map the
    // fail-closed build without conditional source code.
    class UnavailableException : Exception("Trust Wallet Core is unavailable")

    fun generateMnemonic(strength: Int): String {
        val wallet = HDWallet(strength, "")
        return wallet.mnemonic()
    }

    fun isValidMnemonic(mnemonic: String): Boolean = Mnemonic.isValid(mnemonic)
    fun isValidWord(word: String): Boolean = Mnemonic.isValidWord(word)
    fun suggest(prefix: String): List<String> =
        Mnemonic.suggest(prefix).split(" ").filter { it.isNotEmpty() }

    fun entropyFromMnemonic(mnemonic: String): ByteArray {
        if (!isValidMnemonic(mnemonic)) throw InvalidMnemonicException()
        return HDWallet(mnemonic, "").entropy()
    }

    private fun coinType(coin: String): CoinType = when (coin) {
        "eth" -> CoinType.ETHEREUM
        "polygon" -> CoinType.POLYGON
        "base", "arbitrum", "avalanche" -> CoinType.ETHEREUM
        "tron" -> CoinType.TRON
        "solana" -> CoinType.SOLANA
        else -> throw InvalidInputException()
    }

    fun addresses(entropy: ByteArray): Map<String, String> {
        val wallet = HDWallet(entropy, "")
        return mapOf(
            "eth" to wallet.getAddressForCoin(CoinType.ETHEREUM),
            "polygon" to wallet.getAddressForCoin(CoinType.POLYGON),
            "base" to wallet.getAddressForCoin(CoinType.ETHEREUM),
            "arbitrum" to wallet.getAddressForCoin(CoinType.ETHEREUM),
            "avalanche" to wallet.getAddressForCoin(CoinType.ETHEREUM),
            "tron" to wallet.getAddressForCoin(CoinType.TRON),
            "solana" to wallet.getAddressForCoin(CoinType.SOLANA),
        )
    }

    fun publicKeys(entropy: ByteArray): Map<String, ByteArray> {
        val wallet = HDWallet(entropy, "")
        val evm = wallet.getKeyForCoin(CoinType.ETHEREUM)
            .getPublicKeySecp256k1(false).data()
        val tron = wallet.getKeyForCoin(CoinType.TRON)
            .getPublicKeySecp256k1(false).data()
        val solana = wallet.getKeyForCoin(CoinType.SOLANA)
            .getPublicKeyEd25519().data()
        return mapOf(
            "eth" to evm,
            "polygon" to evm,
            "base" to evm,
            "arbitrum" to evm,
            "avalanche" to evm,
            "tron" to tron,
            "solana" to solana,
        )
    }

    fun exportMnemonic(entropy: ByteArray): String = HDWallet(entropy, "").mnemonic()

    data class Signed(val signedTx: ByteArray, val txHash: String)

    fun sign(entropy: ByteArray, coin: String, signingInput: ByteArray): Signed {
        val type = coinType(coin)
        if (type == CoinType.TRON) return signTron(entropy, signingInput)
        if (type == CoinType.SOLANA) return signSolana(entropy, signingInput)
        if (type != CoinType.ETHEREUM && type != CoinType.POLYGON) throw SignFailedException()
        val tx = decodeEip1559(signingInput)
        val key = HDWallet(entropy, "").getKeyForCoin(type)
        val keyBytes = key.data()
        try {
            val transaction = Ethereum.Transaction.newBuilder()
            if (tx.data.isEmpty()) {
                transaction.transfer = Ethereum.Transaction.Transfer.newBuilder()
                    .setAmount(ByteString.copyFrom(tx.value))
                    .build()
            } else {
                transaction.contractGeneric =
                    Ethereum.Transaction.ContractGeneric.newBuilder()
                        .setAmount(ByteString.copyFrom(tx.value))
                        .setData(ByteString.copyFrom(tx.data))
                        .build()
            }
            val input = Ethereum.SigningInput.newBuilder()
                .setChainId(ByteString.copyFrom(tx.chainId))
                .setNonce(ByteString.copyFrom(tx.nonce))
                .setTxMode(Ethereum.TransactionMode.Enveloped)
                .setMaxInclusionFeePerGas(ByteString.copyFrom(tx.maxPriorityFeePerGas))
                .setMaxFeePerGas(ByteString.copyFrom(tx.maxFeePerGas))
                .setGasLimit(ByteString.copyFrom(tx.gasLimit))
                .setToAddress("0x${tx.to.toHex()}")
                .setTransaction(transaction)
                .setPrivateKey(ByteString.copyFrom(keyBytes))
                .build()
            val output = AnySigner.sign(
                input,
                type,
                Ethereum.SigningOutput.parser(),
            )
            if (output.errorValue != 0 || output.encoded.isEmpty) {
                throw SignFailedException()
            }
            val encoded = output.encoded.toByteArray()
            return Signed(encoded, "0x${Hash.keccak256(encoded).toHex()}")
        } catch (e: SignFailedException) {
            throw e
        } catch (_: Exception) {
            throw SignFailedException()
        } finally {
            keyBytes.fill(0)
        }
    }

    /** Signs canonical TRON Transaction.raw bytes and emits broadcasthex JSON. */
    private fun signTron(entropy: ByteArray, rawData: ByteArray): Signed {
        if (rawData.isEmpty()) throw InvalidInputException()
        val key = HDWallet(entropy, "").getKeyForCoin(CoinType.TRON)
        val keyBytes = key.data()
        try {
            val txIdBytes = Hash.sha256(rawData)
            val signature = key.sign(txIdBytes, Curve.SECP256K1)
            if (signature.size != 65) throw SignFailedException()
            val txId = txIdBytes.toHex()
            val json =
                """{"raw_data_hex":"${rawData.toHex()}","signature":["${signature.toHex()}"],"txID":"$txId"}"""
                    .toByteArray(Charsets.UTF_8)
            return Signed(json, txId)
        } catch (e: SignFailedException) {
            throw e
        } catch (_: Exception) {
            throw SignFailedException()
        } finally {
            keyBytes.fill(0)
        }
    }

    /** Signs a legacy Solana message and emits the canonical wire transaction. */
    private fun signSolana(entropy: ByteArray, message: ByteArray): Signed {
        if (message.isEmpty()) throw InvalidInputException()
        val key = HDWallet(entropy, "").getKeyForCoin(CoinType.SOLANA)
        val keyBytes = key.data()
        try {
            val signature = key.sign(message, Curve.ED25519)
            if (signature.size != 64) throw SignFailedException()
            val encoded = byteArrayOf(1) + signature + message
            return Signed(encoded, Base58.encodeNoCheck(signature))
        } catch (e: SignFailedException) {
            throw e
        } catch (_: Exception) {
            throw SignFailedException()
        } finally {
            keyBytes.fill(0)
        }
    }

    private data class Eip1559Fields(
        val chainId: ByteArray,
        val nonce: ByteArray,
        val maxPriorityFeePerGas: ByteArray,
        val maxFeePerGas: ByteArray,
        val gasLimit: ByteArray,
        val to: ByteArray,
        val value: ByteArray,
        val data: ByteArray,
    )

    private data class RlpItem(
        val bytes: ByteArray,
        val isList: Boolean,
        val next: Int,
    )

    /** Decodes the exact unsigned type-2 envelope emitted by `chains`. */
    private fun decodeEip1559(input: ByteArray): Eip1559Fields {
        if (input.size < 3 || input[0] != 0x02.toByte()) {
            throw InvalidInputException()
        }
        val outer = readRlp(input, 1)
        if (!outer.isList || outer.next != input.size) throw InvalidInputException()
        val fields = mutableListOf<RlpItem>()
        var offset = 0
        while (offset < outer.bytes.size) {
            val item = readRlp(outer.bytes, offset)
            fields += item
            offset = item.next
        }
        if (fields.size != 9 || fields.take(8).any { it.isList }) {
            throw InvalidInputException()
        }
        // V1 only permits an empty access list.
        if (!fields[8].isList || fields[8].bytes.isNotEmpty()) {
            throw InvalidInputException()
        }
        val to = fields[5].bytes
        if (to.size != 20) throw InvalidInputException()
        return Eip1559Fields(
            fields[0].bytes,
            fields[1].bytes,
            fields[2].bytes,
            fields[3].bytes,
            fields[4].bytes,
            to,
            fields[6].bytes,
            fields[7].bytes,
        )
    }

    private fun readRlp(source: ByteArray, start: Int): RlpItem {
        if (start !in source.indices) throw InvalidInputException()
        val prefix = source[start].toInt() and 0xff
        return when {
            prefix <= 0x7f -> RlpItem(byteArrayOf(source[start]), false, start + 1)
            prefix <= 0xb7 -> sliceRlp(source, start, prefix - 0x80, 1, false)
            prefix <= 0xbf -> {
                val sizeBytes = prefix - 0xb7
                sliceRlp(source, start, readLength(source, start + 1, sizeBytes),
                    1 + sizeBytes, false)
            }
            prefix <= 0xf7 -> sliceRlp(source, start, prefix - 0xc0, 1, true)
            else -> {
                val sizeBytes = prefix - 0xf7
                sliceRlp(source, start, readLength(source, start + 1, sizeBytes),
                    1 + sizeBytes, true)
            }
        }
    }

    private fun sliceRlp(
        source: ByteArray,
        start: Int,
        length: Int,
        header: Int,
        isList: Boolean,
    ): RlpItem {
        val from = start + header
        val end = from + length
        if (length < 0 || from < 0 || end < from || end > source.size) {
            throw InvalidInputException()
        }
        return RlpItem(source.copyOfRange(from, end), isList, end)
    }

    private fun readLength(source: ByteArray, start: Int, count: Int): Int {
        if (count !in 1..4 || start + count > source.size || source[start] == 0.toByte()) {
            throw InvalidInputException()
        }
        var value = 0
        repeat(count) { value = (value shl 8) or (source[start + it].toInt() and 0xff) }
        return value
    }

    private fun ByteArray.toHex(): String =
        joinToString("") { "%02x".format(it.toInt() and 0xff) }
}
